### PR TITLE
BUG: preserve object dtype for Index set ops

### PR DESCRIPTION
**Pull request by jbrockmendel**
I'd be fine with coercing if we did it consistently, but ATM we behave differently for monotonic-increasing vs otherwise.

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -1047,6 +1047,26 @@ def test_setops_disallow_true(self, method):
         with pytest.raises(ValueError, match="The 'sort' keyword only takes"):
             getattr(idx1, method)(idx2, sort=True)
 
+    def test_setops_preserve_object_dtype(self):
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: k, we should try to split this out the setops to a separate test file (IIRC there is another PR doing of some of this)

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -1047,6 +1047,26 @@ def test_setops_disallow_true(self, method):
...
+        result = idx.intersection(idx[1:])
+        expected = idx[1:]
+        tm.assert_index_equal(result, expected)
+
+        # if other is not monotonic increasing, intersection goes through
+        #  a different route
+        result = idx.intersection(idx[1:][::-1])
+        tm.assert_index_equal(result, expected)
+
+        result = idx._union(idx[1:], sort=None)
```
Comment (jreback) [inline pandas/tests/indexes/test_base.py]: why are you not testing the public method?

**Review by jreback — CHANGES_REQUESTED** [review]
minor comment

File: pandas/tests/indexes/test_base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/test_base.py]: agreed.  i have a local branch separating some of this into tests/indexes/object/... will push it before long

File: pandas/tests/indexes/test_base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/tests/indexes/test_base.py]: we could test that too, this is just "closer" to the bug

Comment (jbrockmendel) [reply pandas/tests/indexes/test_base.py]: updated the test to test the public method as well as the private one

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -2566,6 +2567,7 @@ def _union(self, other, sort):
                 # worth making this faster? a very unusual case
                 value_set = set(lvals)
                 result.extend([x for x in rvals if x not in value_set])
+                result = Index(result)._values  # do type inference here
```
Comment (jreback) [inline pandas/core/indexes/base.py]: why does _wrap_setop_result not do this?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: ATM it does in a roundabout manner.  Index._wrap_setop_result calls self._constructor, which is `type(self)` whereas the subclasses use _shallow_copy.

Comment (jreback) [reply pandas/core/indexes/base.py]: ok

**Review by jreback — APPROVED** [review]

**Pull request merged**